src/server/db_slice.cc

```diff
@@ -1338,4 +1365,49 @@ void DbSlice::ResetUpdateEvents() {
   events_.update = 0;
 }
 
+void DbSlice::TrackKeys(const facade::Connection::WeakRef& conn, const ArgSlice& keys) {
+  if (conn.IsExpired()) {
+    DVLOG(2) << "Connection expired, exiting TrackKey function.";
+    return;
+  }
+
+  DVLOG(2) << "Start tracking keys for client ID: " << conn.GetClientId()
+           << " with thread ID: " << conn.Thread();
+  for (auto key : keys) {
+    DVLOG(2) << "Inserting client ID " << conn.GetClientId()
+             << " into the tracking client set of key " << key;
+    client_tracking_map_[key].insert(conn);
+  }
+}
+
+void DbSlice::SendInvalidationTrackingMessage(const std::string_view& key) {
+  if (client_tracking_map_.find(key) != client_tracking_map_.end()) {
+    // notify all the clients.
+    auto& client_set = client_tracking_map_[key];
+    DVLOG(2) << "Garbage collect clients that are no longer tracking... ";
+    auto is_closed_or_not_tracking = [](const facade::Connection::WeakRef& p) {
+      return (p.IsExpired() || (!p.Get()->IsTrackingOn()));
+    };
+    absl::erase_if(client_set, is_closed_or_not_tracking);
+    DVLOG(2) << "Number of clients left: " << client_set.size();
+
+    if (!client_set.empty()) {
+      auto cb = [key, client_set](unsigned idx, util::ProactorBase*) {
+        for (auto it = client_set.begin(); it != client_set.end(); ++it) {
+          if (it->IsExpired())
+            continue;
+          if ((unsigned int)it->Thread() != idx)
+            continue;
+          facade::Connection* conn = it->Get();
+          conn->SendInvalidationMessageAsync({key});
+          return;
+        }
+      };
+      shard_set->pool()->DispatchBrief(std::move(cb));
+    }
+    // remove this key from the tracking table as the key no longer exists
+    client_tracking_map_.erase(key);
```

Comment:
I'm a little confused why this happens. So after the connection got an invalidate it has read the key again to become subscibed again?

Reply:
correct, this is what Redis' implementation is doing.